src/annotator/integrations/html-side-by-side.js

```diff
@@ -0,0 +1,199 @@
+  const rightMarginVotes = new Map();
+
+  // Gather data about the paragraphs of text in the document.
+  //
+  // In future we might want to expand this to consider other text containers,
+  // since some pages, especially eg. in ebooks, may not have any paragraphs
+  // (eg. instead they may only contain tables or lists or headings).
+  const paragraphs = Array.from(root.querySelectorAll('p'))
+    .map(p => {
+      // Gather some data about them.
+      const rect = p.getBoundingClientRect();
+      const textLength = /** @type {string} */ (p.textContent).length;
+      return { rect, textLength };
+    })
+    .filter(({ rect }) => {
+      // Filter out hidden paragraphs
+      return rect.width > 0 && rect.height > 0;
+    })
+    // Select the paragraphs containing the most text.
+    .sort((a, b) => b.textLength - a.textLength)
+    .slice(0, 15);
+
+  // Let these paragraphs "vote" for what the left and right margins of the
+  // main content area in the document are.
+  paragraphs.forEach(({ rect }) => {
+    let leftVotes = leftMarginVotes.get(rect.left) ?? 0;
+    leftVotes += 1;
+    leftMarginVotes.set(rect.left, leftVotes);
+
+    let rightVotes = rightMarginVotes.get(rect.right) ?? 0;
+    rightVotes += 1;
+    rightMarginVotes.set(rect.right, rightVotes);
+  });
+
+  // Find the margin values with the most votes.
+  if (leftMarginVotes.size === 0 || rightMarginVotes.size === 0) {
+    return null;
+  }
+
+  const leftMargin = [...leftMarginVotes.entries()].sort((a, b) => b[1] - a[1]);
+  const rightMargin = [...rightMarginVotes.entries()].sort(
+    (a, b) => b[1] - a[1]
+  );
+
+  const [leftPos] = leftMargin[0];
+  const [rightPos] = rightMargin[0];
+
+  return { left: leftPos, right: rightPos };
+}
+
+/** @type {Range} */
+let textRectRange;
+
+/**
+ * Return the viewport-relative rect occupied by part of a text node.
+ *
+ * @param {Text} text
+ * @param {number} start
+ * @param {number} end
+ */
+function textRect(text, start = 0, end = text.data.length) {
+  if (!textRectRange) {
+    // Allocate a range only on the first call to avoid the overhead of
+    // constructing and maintaining a large number of live ranges.
+    textRectRange = document.createRange();
+  }
+  textRectRange.setStart(text, start);
+  textRectRange.setEnd(text, end);
+  return textRectRange.getBoundingClientRect();
+}
+
+/**
+ * Yield all the text node descendants of `root` that intersect `rect`.
+ *
+ * @param {Element} root
+ * @param {DOMRect} rect
+ * @return {Generator<Text>}
+ */
+function* textNodesInRect(root, rect) {
+  /** @type {Node|null} */
+  let node = root.firstChild;
+  while (node) {
+    if (node.nodeType === Node.ELEMENT_NODE) {
+      const element = /** @type {Element} */ (node);
+      const elementRect = element.getBoundingClientRect();
+
+      // Skip over subtrees which are entirely outside the viewport or hidden.
+      if (rectIntersects(elementRect, rect)) {
+        yield* textNodesInRect(element, rect);
```

Comment:
How deep can we get here? Are we going to get ourselves in trouble? I suppose every DOM has a finite limit of depth :) ...

Reply:
>  I suppose every DOM has a finite limit of depth :) ...

Indeed the code relies on this assumption. If the DOM tree is absurdly deep (like say thousands of levels or something) I think that would create other problems in the browser before it causes a problem for us.